rfc/rfc-0005-support-composing-supply-chains.md

```diff
@@ -1,3 +1,252 @@
+templates to be stamped out (similar to a supply chain).
+
+## Motivation
+
+At our favorite company, there is a team that is expert at building images.
+They know exactly how each team should do image creation, and that process
+takes a couple of steps of a supply chain. The image building team wants to
+write a master image building supply chain. Every ops team in the company will
+write a supply chain that starts with their own definitions (get code, test, etc),
+then calls out to the image building team’s supply chain to build the image,
+then comes back to the ops team’s supply chain for final steps.
+
+### Implications/Assumptions
+The image-building team does not expect their piece of the supply-chain to
+reconcile with workloads on its own. They are writing a snippet that will only
+make sense in the context of another supply-chain because:
+- If the image-building team's supply-chain reconciles directly with supply-chains,
+  then that supply-chain must have the same selector as any other supply-chain that
+  wants to use it. That doesn't sound like it fits the use case.
+- Presumably the image-building team's supply-chain will be deployed earlier than
+  other supply-chains relying on it. If this early supply-chain tries to reconcile
+  with a workload directly, but the components expect to consume some composed
+  component output, the image-building team's supply-chain will error out immediately.
+
+If the supply-chain isn't reconciling with a workload, it's not really a supply-chain.
+It's some supply-chain snippet.
+
+## Possible Solutions
+
+Kontinue should provide a SupplyChainSnippet CRD.
+
+A supply chain will refer to a snippet by reference (as if the snippet were a
+*Template).
+
+The snippet will define components, just as a supply chain does. A snippet will
+be written with the assumption that it is called by a SupplyChain. The snippet
+will be able to consume values from components in the SupplyChain as if they were
+components in the snippet. SupplyChainSnippet should have no selector.
+
+The snippet will emit some value/type, that would then be available to components
+in the SupplyChain that called the snippet. There are a few way the emitted value
+could be determined:
+- By convention, the last component listed in a Snippet could be the component
+  emitting values to components in the SupplyChain.
+- The Snippet could explicitly define which component could emit values.
+
+_Could a Snippet emit multiple values?_ It should not. If a snippet emitted values
+from two ImageTemplates, for example, there would be no way for the calling
+SupplyChain to differentiate these two values (as the SupplyChain treats the Snippet
+as just another template).
+
+_Should snippets be typed?_ This would result in having a SupplyChainSourceSnippet,
+SupplyChainImageSnippet, SupplyChainConfigSnippet, etc. This multiplication of
+resources adds cognitive and code overhead in exchange for dubious value.
```

Comment:
@scothis Seems like you're in favour of the (option 3) from this list: https://github.com/vmware-tanzu/cartographer/pull/72#issuecomment-1036604300

I'm quite opposed to this option because it blurs the line between a supply chain and a template. I'm much more in favour of adding types to supply chains (option 1)

Reply:
> Seems like you're in favour of the (option 3) from this list: [#72 (comment)](https://github.com/vmware-tanzu/cartographer/pull/72#issuecomment-1036604300)

Mostly, I wouldn't also add support for nested inlines templates.

> I'm quite opposed to this option because it blurs the line between a supply chain and a template. I'm much more in favour of adding types to supply chains (option 1)

What I like about option 3 is that it's a more pure encapsulation. The outward facing contract for a `ClusterXTemplate` from the perspective of a ClusterSupplyChain author is identical, the internal implementation decomposes to multiple sub-components. The main difference is that option 1 forces the supply chain author to know more about the implementation detail of the template.

If the encapsulation of option 3 is leaky, then option 1 is better. But if the encapsulation is complete, option 3 is better. The main place where I see the encapsulation being violated is when tracing resources at runtime. For debugability, the snippet shouldn't be a black box. Tools that visualize the supply chain behavior can choose whether to collapse/expand the internals. The purpose of tracing and debug tools is to violate that encapsulation, so I don't think it's enough of a reason on its own for 1 over 3, but it's something to consider.